### DISCUSSION TITLE
attention implementation on knowledge graph

### DISCUSSION THREAD

**Discussion by junkangwu** [pyg-team/pytorch_geometric#2397] [Q&A]
Hi, I'd like to implement the function as follows:
![image](https://user-images.githubusercontent.com/37494411/114428204-7ad17080-9bee-11eb-97ef-e3652460de1c.png)
![image](https://user-images.githubusercontent.com/37494411/114428240-845ad880-9bee-11eb-88d7-04eddd4634c5.png)
That is, I need to calculate the attention `alpha` between the head entity and relations, and need to calculate the attention `belta` between relation and tail entity.
So, How can I use `pyg` to implement attention score in terms of a part of neighbors, not all neighbors?
Note: I can implement it in the numerator, but it's not clear how do I compute the denominator
![image](https://user-images.githubusercontent.com/37494411/114429542-144d5200-9bf0-11eb-888d-5dfa69765a21.png)
![image](https://user-images.githubusercontent.com/37494411/114429725-4363c380-9bf0-11eb-995d-a0be200794f5.png)
The above is the explanation of the Dimension of the tensor in the softmax function where num_edge is the number of all edge_index.
The calculation of attention is shown below：
![image](https://user-images.githubusercontent.com/37494411/114487302-f3155180-9c41-11eb-8ebc-a44ac407f6b6.png)

Looking forward to your help! Thanks a lot!

**Comment by rusty1s** [comment]
Please see my answer [here](https://github.com/rusty1s/pytorch_geometric/issues/2395).

**Reply by junkangwu** [reply]
@rusty1s  By the way, I find it difficult to compute `alpha` and `beta` at the same time. If I filter `index` via `edge_type`, that is I focus on the subgraph about `relation_i`. Then `beta` attention could be implemented while `alpha` is still hard to address.
For example, if I mask edge_type `i`, then in the `self.propagate`, the `edge_index` is all about `r_i`. So as for attention `alpha`,  I couldn't know the other relations which are the neighbors about entity `u`.

The desired final result shows below:
For example:
`attention_{e_1,e_2} = alpha_{e_1,r_1} * beta_{r_1,e_2} `
![image](https://user-images.githubusercontent.com/37494411/114568650-e545e780-9ca6-11eb-81a5-26d00ac78b3b.png)
Looking forward to your help again!

**Reply by rusty1s** [reply]
See https://github.com/rusty1s/pytorch_geometric/issues/2395
